### DISCUSSION TITLE
What is the state of `tmrl` for Linux ?

### DISCUSSION THREAD

**Discussion by vcharraut** [trackmania-rl/tmrl#39] [Q&A]
Hey!

I'm trying to use `tmrl` to train AIs in TrackMania. So far I got Trackmania 2020 and OpenPlanet working but when I'm trying to play with a pre-trained agents, it says it is only supported with Windows. 

So I was wondering what are in general the limits the package have for development in Linux ? Is it possible to train our own agent without the need of gamepad ?

**Comment by yannbouteiller** [comment]
Hi! It is totally possible in theory, but that will require some work because atm the TrackMania support is only implemented for Windows.

I did not know that TrackMania was available on Linux at all actually. It would be a nice addition to `tmrl` to port compatibility of the TrackMania environment to Linux :) it is probably not very hard either, the hardest (and most interesting for everyone) would be to port `vgamepad` compatibility to Linux, but until someone hopefully takes on the work of doing this someday, the easiest solution is to just use keystrokes, which is already implemented in the TrackMania environment btw I believe. We just never ported this TrackMania environment to Linux, but if you have some skills in python it should be fairly straightforward to do.

**Reply by yannbouteiller** [reply]
Well, not so straightforward actually, because we rely on the win32ui API to handle the game window and capture screenshots efficiently, but still doesn't sound like an impossible task :)

**Reply by vcharraut** [reply]
Thank you for your answer.

I will try to look to create a port for Linux because I'm very interested to play with it! I'll try to take time for it, hope it will not be too complicated

**Reply by yannbouteiller** [reply]
Apparently it is possible to run OpenPlanet on Linux : https://openplanet.dev/docs/help/linux

So, most likely the only potentially difficult part will be to find a good replacement for the `win32ui` library (and code a virtual gamepad with `uinput` probably, if you want a virtual gamepad)

**Reply by yannbouteiller** [reply]
Note: `tmrl` works on Linux, I use Linux all the time for hosting my trainer. It is only the TrackMania environment that you will need to port.

**Reply by vcharraut** [reply]
Yes, I got to install everything with Lutris (TrackMania + OpenPlanet), could install the map and put the game on the top right of the screen but when I launched `python -m tmrl --test`, I got blocked by an assert that is checking is the os is Windows or not

I understand that the gamepad is made to work with Windows, so I was wondering if there were other things that could raise issues.
Futhermore, is there any way to launch a test without the need a virtual gamepad ?

**Reply by yannbouteiller** [reply]
I suggest you read the code of the [TM2020Interface](https://github.com/trackmania-rl/tmrl/blob/master/tmrl/custom/custom_gym_interfaces.py) class, there is a `gamepad` option that you can turn off (you can set this to `false` in `config.json`), then it shouldn't try to use `vgamepad` which is compatible with Windows only.

**Reply by yannbouteiller** [reply]
However, the [window interface](https://github.com/trackmania-rl/tmrl/blob/master/tmrl/custom/utils/window.py) will not work on Linux, so you will have to adapt this.
